### PR TITLE
Implementing tuple access via desugaring and indexing.

### LINKED ISSUE #632: Implement tuple access syntax via destructuring and indexing

### PR DESCRIPTION
**Pull request by emilyaherbert**
Closes #632

### CODE REVIEW

File: sway-core/Cargo.toml [hunk 1]
```diff
@@ -29,6 +29,7 @@ sway-ir = { version = "0.3.1", path = "../sway-ir" }
 sway-types = { version = "0.3.1", path = "../sway-types" }
 generational-arena = "0.2"
 thiserror = "1.0"
+nanoid = "0.4.0"
```
Comment (adlerjohn) [inline sway-core/Cargo.toml:32]: sort pls. Maybe also use `"0.4"` instead of `"0.4.0"`?

File: sway-core/src/build_config.rs [hunk 2]
```diff
@@ -10,6 +13,7 @@ pub struct BuildConfig {
     pub(crate) print_intermediate_asm: bool,
     pub(crate) print_finalized_asm: bool,
     pub(crate) print_ir: bool,
+    pub(crate) generated_names: Arc<Mutex<Vec<&'static str>>>,
```
Comment (sezna) [inline sway-core/src/build_config.rs:16]: why this instead of a `Arc<Vec<&'static str>>`?

File: sway-core/src/parse_tree/declaration.rs [hunk 3]
```diff
@@ -137,129 +129,87 @@ impl Declaration {
     pub(crate) fn parse_from_pair(
         decl: Pair<Rule>,
         config: Option<&BuildConfig>,
-    ) -> CompileResult<Self> {
+    ) -> CompileResult<Vec<Self>> {
```
Comment (sezna) [inline sway-core/src/parse_tree/declaration.rs:132]: could you document in the docstring why this now returns a vector and how it relates to desugaring?

File: sway-core/src/parse_tree/declaration/variable.rs [hunk 4]
```diff
@@ -10,3 +18,231 @@ pub struct VariableDeclaration {
     pub body: Expression, // will be codeblock variant
     pub is_mutable: bool,
 }
+
+pub enum VariableDeclarationLHS {
```
Comment (sezna) [inline sway-core/src/parse_tree/declaration/variable.rs:22]: could we provide some documentation for this? As with all fully-`pub` structs -- we currently have pretty good coverage in other areas

File: sway-core/src/parse_tree/declaration/variable.rs [hunk 5]
```diff
@@ -10,3 +18,231 @@ pub struct VariableDeclaration {
...
+                maybe_body = var_decl_parts.next().unwrap();
+                Some(type_asc)
+            }
+            _ => None,
+        };
+        let type_ascription_span = type_ascription.clone().map(|x| Span {
+            span: x.into_inner().next().unwrap().as_span(),
+            path: config.map(|x| x.path()),
+        });
+        let type_ascription = if let Some(ascription) = type_ascription {
```
Comment (sezna) [inline sway-core/src/parse_tree/declaration/variable.rs:69]: I think we should have positive/negative test cases for the following, would you mind adding them?
```rust
// 1
let (x, y): (u32, bool) = (10, true);
// 2
let (x, y): (u32, _) = (42, true);
// 3
fn test<T, E>(a: T, b: E) {
     let (x, y): (T, E) = (a, b);
} 
test(true, false);
// ensure monomorphizing works in tuples by instantiating with different types
test (42, 42);

// 4
let (a, (b, (c, d) ) ): (u64, (u32, (bool, str[2]) ) ) = (42u64, (42u32, (true, "ok") ) );

````

File: test/src/e2e_vm_tests/test_programs/tuple_indexing/Forc.toml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+[project]
+author = "Fuel Labs <contact@fuel.sh>"
+license = "Apache-2.0"
+name = "tuple_indexing"
+entry = "main.sw"
+
+[dependencies]
+std = { git = "http://github.com/FuelLabs/sway-lib-std" }
+core = { git = "http://github.com/FuelLabs/sway-lib-core" }
```
Comment (sezna) [inline test/src/e2e_vm_tests/test_programs/tuple_indexing/Forc.toml:9]:
```suggestion
```

File: test/src/e2e_vm_tests/test_programs/tuple_access/Forc.toml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+[project]
+author = "Fuel Labs <contact@fuel.sh>"
+license = "Apache-2.0"
+name = "tuple_access"
+entry = "main.sw"
+
+[dependencies]
+std = { git = "http://github.com/FuelLabs/sway-lib-std" }
+core = { git = "http://github.com/FuelLabs/sway-lib-core" }
```
Comment (sezna) [inline test/src/e2e_vm_tests/test_programs/tuple_access/Forc.toml:9]:
```suggestion
```

**Review by sezna — CHANGES_REQUESTED** [review]
Overall looks good but I think the test coverage needs to be improved. See my comment for some test cases.

File: sway-core/src/build_config.rs (hunk 2 above)
Comment (emilyaherbert) [reply sway-core/src/build_config.rs:16]: `Arc<Vec<&'static str>>` makes it difficult to push to the vec. There may be another way to accomplish this, but this is the one I am familiar with.

**Review by sezna — APPROVED** [review]
LGTM, awesome!

**Pull request merged**